support django, svector, bvector and vecf16

cutecutecat opened the pull request:
Close #2 Close #3 Close #4
Close https://github.com/tensorchord/pgvecto.rs/issues/518

## Feature
Add supports for:

|                                                        | [Vector](https://docs.pgvecto.rs/usage/indexing.html) | [Sparse Vector](https://docs.pgvecto.rs/reference/vector-types/svector.html) | [Half-Precision Vector](https://docs.pgvecto.rs/reference/vector-types/vecf16.html) | [Binary Vector](https://docs.pgvecto.rs/reference/vector-types/bvector.html) |
| ------------------------------------------------------ | ----------------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy) | ✅Insert                                               | ✅Insert                                                      | ✅Insert                                                      | ✅Insert                                                      |
| [Psycopg3](https://github.com/psycopg/psycopg)         | ✅Insert ✅Copy                                         | ✅Insert ✅Copy                                                | ✅Insert ✅Copy                                                | ✅Insert ✅Copy                                                |
| [Django](https://github.com/django/django)             | ✅Insert                                               | ✅Insert                                                      | ✅Insert                                                      | ✅Insert                                                      |

## Usages
 Now we can build a sparse vector from these statements:
```python
import numpy as np
from scipy.sparse import coo_array
from pgvecto_rs.types import SparseVector

SparseVector.from_parts(3, [0, 2], [1.0, 3.0])
SparseVector({0: 1.0, 1: 2.0, 2: 3.0}, 3)
SparseVector(coo_array((np.array([2.0, 3.0]), np.array([[1, 2]])), shape=(3,)))
SparseVector(coo_array(np.array[1, 0, 2, 0, 3, 0, 0, 0])
```

cutecutecat committed:
support django, svector, bvector and vecf16

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

kemingy commented on README.md:
```diff
@@ -2,23 +2,33 @@
 <a href="https://pypi.org/project/pgvecto_rs/"><img src="https://img.shields.io/pypi/dm/pgvecto_rs.svg?label=Pypi%20downloads" alt="trackgit-views" /></a>
 </p>
 
+[PGVecto.rs](https://github.com/tensorchord/pgvecto.rs) support for Python, supports Django, SQLAlchemy and Psycopg 3.
```
```suggestion
[PGVecto.rs](https://github.com/tensorchord/pgvecto.rs) Python library, supports Django, SQLAlchemy, and Psycopg 3.
```

cutecutecat replied:
Fixed.

cutecutecat committed:
fix: quantization parse error

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

kemingy commented on pyproject.toml:
```diff
@@ -7,15 +7,15 @@ authors = [
 license = { text = "Apache-2.0" }
 classifiers = [
     "Programming Language :: Python :: 3 :: Only",
-    "Programming Language :: Python :: 3.8",
```
Does py3.8 have any conflicts? The EoS is this Oct.

cutecutecat replied:
I cannot find a version of scipy for both `3.8` and `3.12`:
scipy 1.10 >=3.8,<3.12
scipy 1.11 >=3.9

kemingy commented on src/pgvecto_rs/types/bvector.py:
```diff
@@ -0,0 +1,107 @@
+    for i, x in enumerate(packed_arr):
+        for j in range(64):
+            unpacked_arr[i * 64 + j] = (x >> j) & 1
+    return unpacked_arr
```
You can use `np.packbits(arr).view(np.uint64)`

kemingy commented on src/pgvecto_rs/types/index.py:
```diff
@@ -0,0 +1,91 @@
+        threads: Optional[int] = None,
+    ):
+        self.index = index
+        self.threads = threads or os.cpu_count() or 1
```
The CPU count here is the client, not the Pg server.

cutecutecat replied:
Yes, you are right, threads auto filling at client is removed.

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,207 @@
+    SparseShapeError,
+)
+
+NO_DEFAULT = object()
```
Defining a dummy class would be better. You can refer to `dataclasses.MISSING`

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,207 @@
+        return vec
+
+    def to_numpy(self):
+        vec = np.repeat(0.0, self._dim).astype(np.float32)
```
Use `np.zeros`

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,207 @@
+
+    def to_numpy(self):
+        vec = np.repeat(0.0, self._dim).astype(np.float32)
+        for i, v in zip(self._indices, self._values):
+            vec[i] = v
```
```suggestion
        vec[self._indices] = self._values
```

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,207 @@
+
+    def _from_dense(self, value):
+        self._dim = len(value)
+        self._indices = [i for i, v in enumerate(value) if v != 0]
```
I guess `v` is float, so it's not a good idea to compare it with 0 directly. Try `math.isclose` or `np.isclose`.

kemingy reviewed (commented):
It's a huge PR. I only reviewed the type part. Will review others later.

zbloss commented:
For what it's worth, I have tried pulling down these changes and am having serialization issues with `tensorchord/pgvecto-rs:pg16-v0.3.0`

`ValueError: Cannot serialize: <pgvecto_rs.types.index.IndexOption object at 0xffff5fa9a500>`

cutecutecat committed:
fix by comment and refactor django.indexes

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

cutecutecat commented:
> For what it's worth, I have tried pulling down these changes and am having serialization issues with `tensorchord/pgvecto-rs:pg16-v0.3.0`
> 
> `ValueError: Cannot serialize: <pgvecto_rs.types.index.IndexOption object at 0xffff5fa9a500>`

@zbloss Thanks for your report, I found that Django cannot serialize a customed struct, only basic types, so I changed Index to a flatten struct, like pgvector-python does.

The tests I have implemented before are failed to catch it.
 If convenient, can you share me your codes or workflow to let me catch this case in future tests?

zbloss commented:
Thanks for looking at this quickly. I'm now getting a new error when running `manage.py makemigrations` in my project 

`ImportError: cannot import name 'Index' from 'pgvecto_rs.django' (/usr/local/lib/python3.10/site-packages/pgvecto_rs/django/__init__.py)`

I can't share the code, but it is a fairly simple django project. This error occurs when trying to `makemigrations`.

cutecutecat commented:
> Thanks for looking at this quickly. I'm now getting a new error when running `manage.py makemigrations` in my project
> 
> `ImportError: cannot import name 'Index' from 'pgvecto_rs.django' (/usr/local/lib/python3.10/site-packages/pgvecto_rs/django/__init__.py)`
> 
> I can't share the code, but it is a fairly simple django project. This error occurs when trying to `makemigrations`.

Sorry about that, as this PR is under heavy developing, the API may changes rapidly before fully reviewed and tested. After today's refactor, there is no `Index` at `pgvecto_rs.django` now, instead we have `HnswIndex`, `IvfIndex` and `FlatIndex`:

```python
from django.db import models
from pgvecto_rs.django import HnswIndex, IvfIndex
from pgvecto_rs.types import IndexOption, Hnsw


class Item(models.Model):
    class Meta:
        indexes = [
            HnswIndex(
                name="emb_idx_1",
                fields=["embedding"],
                opclasses=["vector_l2_ops"],
                # don't pass any of `m`, `ef_construction`, `threads`, `quantization_type` or `quantization_ratio`
                # if created by `with_option`, they will be overwritten
            ).with_option(
                IndexOption(index=Hnsw(m=16, ef_construction=100), threads=1)
            ),
            # or
            IvfIndex(
                name="emb_idx_2",
                fields=["embedding"],
                nlist=3,
                opclasses=["vector_l2_ops"],
            ),
        ]
```

cutecutecat committed:
compact with python 3.8

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,214 @@
+        elements, dim = value.split("/", 2)
+        indices = []
+        values = []
+        for e in elements[1:-1].split(","):
```
It's better to use `str.find('{')` and `str.find('}')` in case there are spaces.

kemingy commented on src/pgvecto_rs/types/svector.py:
```diff
@@ -0,0 +1,214 @@
+        indices = np.frombuffer(bytes, dtype="<I", count=length, offset=0).astype(
+            np.uint32
+        )
+        bytes = bytes[4 * length :]
```
Use `memoryview` to avoid copying the underlying bytes.

kemingy commented on src/pgvecto_rs/types/vecf16.py:
```diff
@@ -0,0 +1,86 @@
+from pgvecto_rs.errors import NDArrayDimensionError, ToDBDimUnequalError
+
+
+class Float16Vector:
```
It seems `Float16Vector` is the same as `Vector` except for the dtype part. Can we inherit and modify the class attribute to f16?

kemingy commented on src/pgvecto_rs/django/indexes.py:
```diff
@@ -0,0 +1,173 @@
+        quantization = option.index.quantization
+
+        self.quantization_type = quantization.type if quantization else None
+        self.quantization_ratio = quantization.ratio if quantization else None
```
Please also add some documents to the children's classes' method since it will override all the options, even when the quantization is none.

cutecutecat committed:
fix with comments

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

kemingy commented on src/pgvecto_rs/types/vector.py:
```diff
@@ -0,0 +1,90 @@
+
+        return value.to_binary()
+
+    @classmethod
+    def _from_db(cls, value):
+        if value is None or isinstance(value, np.ndarray):
+            return value
+
+        return cls.from_text(value).to_numpy().astype(np.float32)
+
+    @classmethod
+    def _from_db_binary(cls, value):
+        if value is None or isinstance(value, np.ndarray):
+            return value
+
+        return cls.from_binary(value).to_numpy().astype(np.float32)
```
It seems the `to_numpy().astype(np.float32)` will convert it to np.array, but it's a classmethod.

kemingy commented on README.md:
```diff
@@ -31,9 +41,277 @@ docker run \
+                name="emb_idx_1",
+                fields=["embedding"],
+                opclasses=["vector_l2_ops"],
+                # don't pass any of `m`, `ef_construction`, `threads`, `quantization_type` or `quantization_ratio`
+                # if created by `with_option`, they will be overridden
+            ).with_option(
+                IndexOption(index=Hnsw(m=16, ef_construction=100), threads=1)
+            ),
```
When do we need to call the `with_option`? I don't know if there is such kind of scenario. Only the `threads` can be changed on the fly.

cutecutecat replied:
I think people may want to create Index by a structured `IndexOption` instead of a set of arguments before, but as you say, it introduces more complex. I would remove this usage.

sskorol commented:
I tried this branch on my pet project. Was specifically interested in `SparseVector` implementation. Used it to play with hybrid search described [here](https://docs.pgvecto.rs/use-case/sparse-vector.html). Works good so far with SQLAlchemy. Thanks @cutecutecat!

cutecutecat committed:
change vector from_db into cls instead of ndarray

Signed-off-by: cutecutecat <junyuchen@tensorchord.ai>

cutecutecat commented:
> I tried this branch on my pet project. Was specifically interested in `SparseVector` implementation. Used it to play with hybrid search described [here](https://docs.pgvecto.rs/use-case/sparse-vector.html). Works good so far with SQLAlchemy. Thanks @cutecutecat!

@sskorol I am happy that you are interested in our new sparse features, we can simplify the procession of sparse vectors now, and no need to extract indices and values from model output. I would update that article soon, with our new powerful features in SDK v0.2.0!

sskorol commented:
@cutecutecat thanks! I'm doing it this way for now. But I'm curious to see how it could be done optimally.
```python
class VectorDAO(BaseDAO[VectorRecord, VectorCreate, VectorUpdate]):
    def __init__(
        self,
        model: Type[ModelType],
        session_factory: Callable[..., ContextManager[Session]],
    ):
        super().__init__(model, session_factory)

    def insert_vector(self, vector: VectorCreate):
        with self.session_factory() as db:
            stmt = insert(VectorRecord).values(**vector.model_dump())
            db.execute(stmt)
            db.commit()

    def get_top_matches(
        self, search_vector: VectorSearch, top_k: int = 10
    ) -> List[ModelType]:
        with self.session_factory() as db:
            dense_query = (
                db.query(self.model)
                .order_by(self._l2_distance(search_vector.v_dense))
                .limit(top_k)
                .subquery()
            )
            sparse_query = (
                db.query(self.model)
                .order_by(self._dot_product(search_vector.v_sparse))
                .limit(top_k)
                .subquery()
            )

            dense_alias = aliased(self.model, dense_query)
            sparse_alias = aliased(self.model, sparse_query)
            final_query = db.query(dense_alias).union_all(db.query(sparse_alias))

            return final_query.all()

    def _l2_distance(self, dense_vector: Vector):
        return self.model.v_dense.op("<->")(dense_vector)

    def _dot_product(self, sparse_vector: SparseVector):
        return self.model.v_sparse.op("<#>")(sparse_vector)
```
And index/extension is added via alembic migration:
```python
def upgrade() -> None:
    op.execute("CREATE EXTENSION IF NOT EXISTS vectors")
    op.create_table(
        "vectorrecord",
        # ...
        sa.Column("v_dense", VECTOR(1024), nullable=False),
        sa.Column("v_sparse", SVECTOR(250002), nullable=False),
    )
    op.execute(
        """
            CREATE INDEX IF NOT EXISTS idx_sparse_vector ON vectorrecord
            USING vectors (v_sparse svector_dot_ops)
            WITH (options = '[indexing.hnsw]');
        """
    )
    op.execute(
        """
            CREATE INDEX IF NOT EXISTS idx_dense_vector ON vectorrecord
            USING vectors (v_dense vector_l2_ops)
            WITH (options = '[indexing.hnsw]');
        """
    )
```
